### PR TITLE
JBPM-4646 - Case management

### PR DESCRIPTION
**Pull request by mswiderski**
depends on https://github.com/droolsjbpm/jbpm/pull/558

### CODE REVIEW

**Comment by mswiderski** [comment]
@cristianonicolai @krisv first version ready to be reviewed and merged

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseAdHocFragmentList.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseAdHocFragmentList.java]: 2016

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseCommentList.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseCommentList.java]: 2016

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseDefinitionList.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseDefinitionList.java]: 2016

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseInstanceList.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseInstanceList.java]: 2016

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseMilestoneList.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseMilestoneList.java]: 2016

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseRoleAssignmentList.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseRoleAssignmentList.java]: 2016

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    void addComment(String containerId, String caseId, String author, String text);
+
+    void removeComment(String containerId, String caseId, String commentId);
+
+    List<CaseInstance> getCaseInstances(Integer page, Integer pageSize);
+
+    List<CaseInstance> getCaseInstances(List<Integer> status, Integer page, Integer pageSize);
+
+    List<CaseInstance> getCaseInstancesOwnedBy(String owner, List<Integer> status, Integer page, Integer pageSize);
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java]: I would also suggest to add getCaseInstancesByStatus

**Comment by mswiderski** [comment]
jenkins retest this

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java (hunk 7 above)
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java]: the one above is pretty much getCaseInstancesByStatus it just does not have the ByStatus in method name

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseServiceIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+            // since roles were not assigned to any users/groups no tasks are available
+            List<TaskSummary> tasks = taskClient.findTasksAssignedAsPotentialOwner(USER_YODA, 0, 10);
+            assertEquals(0, tasks.size());
+
+            caseClient.destroyCaseInstance(CONTAINER_ID, caseId);
+
+            try {
+                // this should throw exception as there is no case any more
+                caseClient.getCaseInstance(CONTAINER_ID, caseId);
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-all/src/test/java/org/kie/server/integrationtests/jbpm/cases/CaseServiceIntegrationTest.java]: Adding a fail() in the next line would ensure that the exception was actually thrown in the test.

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java (hunk 7 above)
Comment (cristianonicolai) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java]: True, my bad.

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+    void triggerAdHocFragment(String containerId, String caseId, String adHocName, Map<String, Object> data);
+
+    void triggerAdHocFragmentInStage(String containerId, String caseId, String stageId, String adHocName, Map<String, Object> data);
+
+    List<CaseMilestone> getMilestones(String containerId, String caseId, boolean achievedOnly, Integer page, Integer pageSize);
+
+    List<CaseStage> getStages(String containerId, String caseId, boolean activeOnly, Integer page, Integer pageSize);
+
+    List<CaseAdHocFragment> getAdHocFragments(String containerId, String caseId);
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java]: Would this method consider the stage that the case is in? For instance, certain AdHoc fragment would be only available inside a stage and another fragment is available at any stage in the case.

Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java]: this returns only ad hoc tasks (those that have not incoming connection) that are on the case/process level but not within a stage. Stage ad hoc fragments are returned from getStages()::CaseStage.getAdHocFragments

Comment (cristianonicolai) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/CaseServicesClient.java]: sounds good, thanks

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseDefinition.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+package org.kie.server.api.model.cases;
+
+import java.util.List;
+import java.util.Map;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+import org.kie.server.api.model.instance.NodeInstance;
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/model/cases/CaseDefinition.java]: unused import

**Comment by cristianonicolai** [comment]
@mswiderski this looks really good, just some minor comments to fix.

**Comment by mswiderski** [comment]
@cristianonicolai changes applied, thanks

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/helper/CaseServicesClientBuilder.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/helper/CaseServicesClientBuilder.java]: 2016

File: kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseKieContainerCommandServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (cristianonicolai) [inline kie-server-parent/kie-server-services/kie-server-services-case-mgmt/src/main/java/org/kie/server/services/casemgmt/CaseKieContainerCommandServiceImpl.java]: 2016

**Comment by mswiderski** [comment]
license headers updated, thanks @cristianonicolai

**Comment by cristianonicolai** [comment]
@mswiderski just spotted two test classes without license headers CaseServiceIntegrationTest and
CaseRuntimeDataServiceIntegrationTest.

**Comment by mswiderski** [comment]
added missing license header to test classes and added updateComment endpoint to mirror services capabilities.

**Comment by cristianonicolai** [comment]
@mswiderski one of the failures on jenkins is related to the jbpm-case-mgmt dependency on droolsjbpm-bpms-distribution/pom.xml. I guess that needs to be removed and the new modules added.

**Comment by cristianonicolai** [comment]
droolsjbpm-bpms-distribution/src/main/assembly/pre-bin.xml also has an include to the old module.

**Comment by mswiderski** [comment]
@psiroky could you please take a look at extra commit to fix the build https://github.com/droolsjbpm/droolsjbpm-integration/pull/601/commits/180afd7c88f55dcf40c950bc963b7c3bf09029e3 it's also required on master as the build seems to be failing there...

**Comment by psiroky** [comment]
@mswiderski thanks for noticing that! I've pushed your fix directory to master/6.5.x as this is a hot fix which is needed to make the build work. And sorry for the troubles...

**Comment by mswiderski** [comment]
@psiroky @mariofusco I don't think blueprint test failures are related to this PR but just double checking. Any hints?

**Comment by cristianonicolai** [comment]
@mswiderski the dependencies in here also need some checking https://github.com/droolsjbpm/droolsjbpm-integration/blob/master/droolsjbpm-bpms-distribution/pom.xml#L133, probably removing the old module and adding the two new ones.

**Comment by mswiderski** [comment]
thanks @cristianonicolai @psiroky for helping out, now the build is green :)

**Comment by mswiderski** [comment]
ready to be merged from my point of view

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by cristianonicolai** [comment]
+1

**Pull request merged**